File: packages/object-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/objectdetection/ObjectDetection.java
package io.capawesome.capacitorjs.plugins.mlkit.objectdetection;

import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.objects.ObjectDetector;
import com.google.mlkit.vision.objects.defaults.ObjectDetectorOptions;
import io.capawesome.capacitorjs.plugins.mlkit.objectdetection.classes.options.ProcessImageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.objectdetection.classes.results.ProcessImageResult;
import io.capawesome.capacitorjs.plugins.mlkit.objectdetection.interfaces.NonEmptyResultCallback;

public class ObjectDetection {

    @NonNull
    private final ObjectDetectionPlugin plugin;

    public ObjectDetection(@NonNull ObjectDetectionPlugin plugin) {
        this.plugin = plugin;
    }

    public void processImage(@NonNull ProcessImageOptions options, @NonNull NonEmptyResultCallback<ProcessImageResult> callback) {
        InputImage inputImage = createInputImageFromFilePath(options.getPath());
        if (inputImage == null) {
            callback.error(new Exception(ObjectDetectionPlugin.ERROR_LOAD_IMAGE_FAILED));
            return;
        }

        ObjectDetectorOptions.Builder optionsBuilder = new ObjectDetectorOptions.Builder()
            .setDetectorMode(ObjectDetectorOptions.SINGLE_IMAGE_MODE);
        if (options.isClassificationEnabled()) {
            optionsBuilder.enableClassification();
        }
        if (options.isMultipleObjectsEnabled()) {
            optionsBuilder.enableMultipleObjects();
        }

        final ObjectDetector objectDetector = com.google.mlkit.vision.objects.ObjectDetection.getClient(optionsBuilder.build());
        plugin
            .getActivity()
            .runOnUiThread(() -> {
                objectDetector
                    .process(inputImage)
                    .addOnSuccessListener(detectedObjects -> {
                        objectDetector.close();
                        ProcessImageResult result = new ProcessImageResult(detectedObjects);
                        callback.success(result);
                    })
                    .addOnCanceledListener(() -> {
                        objectDetector.close();
                        callback.error(new Exception(ObjectDetectionPlugin.ERROR_PROCESS_IMAGE_CANCELED));
                    })
                    .addOnFailureListener(exception -> {
                        objectDetector.close();
                        callback.error(exception);
                    });
            });
    }

    @Nullable
    private InputImage createInputImageFromFilePath(@NonNull String path) {
        try {
            return InputImage.fromFilePath(plugin.getContext(), Uri.parse(path));
        } catch (Exception exception) {
            return null;
        }
    }
}

File: packages/object-detection/ios/Plugin/ObjectDetection.swift
import Foundation
import MLKitVision
import MLKitObjectDetection

@objc public class ObjectDetection: NSObject {
    public let plugin: ObjectDetectionPlugin

    init(plugin: ObjectDetectionPlugin) {
        self.plugin = plugin
    }

    @objc func processImage(_ options: ProcessImageOptions, completion: @escaping (ProcessImageResult?, Error?) -> Void) {
        guard let visionImage = createVisionImageFromFilePath(options.getPath()) else {
            completion(nil, CustomError.loadImageFailed)
            return
        }

        let objectDetectorOptions = ObjectDetectorOptions()
        objectDetectorOptions.detectorMode = .singleImage
        objectDetectorOptions.shouldEnableClassification = options.isClassificationEnabled()
        objectDetectorOptions.shouldEnableMultipleObjects = options.isMultipleObjectsEnabled()

        let objectDetector = ObjectDetector.objectDetector(options: objectDetectorOptions)

        do {
            let objects = try objectDetector.results(in: visionImage)
            let result = ProcessImageResult(objects: objects)
            completion(result, nil)
        } catch let error {
            completion(nil, error)
        }
    }

    private func createVisionImageFromFilePath(_ path: String) -> VisionImage? {
        guard let url = URL.init(string: path) else {
            return nil
        }
        if FileManager.default.fileExists(atPath: url.path) {
            guard let image = UIImage.init(contentsOfFile: url.path) else {
                return nil
            }
            return VisionImage.init(image: image)
        } else {
            return nil
        }
    }
}
